### PR TITLE
save boundary and interface material properties on LD elements

### LINKED ISSUE #16: Save boundary and interface properties to LD elements
We should be able to save interface and boundary properties on lower dimensional elements for visualization purposes. This will help a lot when post-processing and visualizing data from CP simulations including CZM models

### PR DESCRIPTION
**Pull request by arovinelli**
This pull requests allows deer to save boundary and interface properties on LD elements.
At the moment  and RealVectorValue material properties can be saved. It will extremely easy to add other types. Templating might be an option but I didn't bother
close #16

### CODE REVIEW

**Comment by arovinelli** [comment]
@lindsayad please take also a look at this and tell us if this is omething you guiys might be interested in.
My idea would be add maps to MooseMesh when generating LD elements with MeshGenerators to avoid duplicates in the UO (maybe with an optional parameter).

**Comment by lindsayad** [comment]
I think it looks interesting, I personally don't use user objects a lot; usually if I think it's generally applicable enough I like to make it into a typical MOOSE system. So conceptually what is this? Stateful lower dimensional elements?

**Comment by arovinelli** [comment]
This is a system that allows to save boundary and interface material property in lower dimensional elements to preserve element side attributes when you have domain triple junctions.
It uses an InterfaceUO to first create the map between elements side and LD elements, then the same interface user object is used to save the properties at each qp over a boundary or interface.The LD restricted auxkernel then probes the UO to save the properties on the LD domain.
It is more of a visualization helper than anything else, but it is somethign yhou really need if you want to do some serious postprocessing on boundary/interface results with complex geometries.



> I think it looks interesting, I personally don't use user objects a lot; usually if I think it's generally applicable enough I like to make it into a typical MOOSE system. So conceptually what is this? Stateful lower dimensional elements?

File: include/userobjects/Map2LDInterfaceMPReal.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+class Map2LDInterfaceMPReal : public Map2LDInterfaceMaterialPropertyUO {
+public:
+  static InputParameters validParams();
+
+  Map2LDInterfaceMPReal(const InputParameters &parameters);
+  virtual ~Map2LDInterfaceMPReal();
+
+protected:
+  virtual Real getMPValue(const unsigned int qp) const override {
```
Comment (reverendbedford) [inline include/userobjects/Map2LDInterfaceMPReal.h]: Does this need to be inlined?

File: include/userobjects/Map2LDInterfaceMPReal.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#pragma once
+
+#include "Map2LDInterfaceMaterialPropertyUO.h"
+
+class Map2LDInterfaceMPReal;
+
+template <> InputParameters validParams<Map2LDInterfaceMPReal>();
+
+/**
```
Comment (reverendbedford) [inline include/userobjects/Map2LDInterfaceMPReal.h]: I mean either include a comment or don't :)

File: include/userobjects/Map2LDInterfaceMPRealVectorValue.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#pragma once
+
+#include "Map2LDInterfaceMaterialPropertyUO.h"
+
+class Map2LDInterfaceMPRealVectorValue;
+
+template <> InputParameters validParams<Map2LDInterfaceMPRealVectorValue>();
+
+/**
```
Comment (reverendbedford) [inline include/userobjects/Map2LDInterfaceMPRealVectorValue.h]: Comment?

File: include/userobjects/Map2LDInterfaceMaterialPropertyUO.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+#pragma once
+
+#include "Map2LDelem.h"
+
+class Map2LDInterfaceMaterialPropertyUO;
+
+template <> InputParameters validParams<Map2LDInterfaceMaterialPropertyUO>();
+
+/**
```
Comment (reverendbedford) [inline include/userobjects/Map2LDInterfaceMaterialPropertyUO.h]: ...

File: include/userobjects/Map2LDelem.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+#pragma once
+
+#include "InterfaceUserObject.h"
+
+class Map2LDelem;
+
+template <> InputParameters validParams<Map2LDelem>();
+
+/**
```
Comment (reverendbedford) [inline include/userobjects/Map2LDelem.h]: ...

File: include/userobjects/Map2LDelem.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  static InputParameters validParams();
+
+  Map2LDelem(const InputParameters &parameters);
+  virtual ~Map2LDelem();
+  virtual void initialSetup() override;
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override { return; };
+  virtual void threadJoin(const UserObject & /*uo*/) override { return; };
+  // Real getQpValueForLD(dof_id_type ld_elem, unsigned int qp) const;
```
Comment (reverendbedford) [inline include/userobjects/Map2LDelem.h]: Well this shouldn't still be here.

File: include/userobjects/Map2LDelem.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+public:
+  static InputParameters validParams();
+
+  Map2LDelem(const InputParameters &parameters);
+  virtual ~Map2LDelem();
+  virtual void initialSetup() override;
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override { return; };
+  virtual void threadJoin(const UserObject & /*uo*/) override { return; };
```
Comment (reverendbedford) [inline include/userobjects/Map2LDelem.h]: Oh god, why the comment out?

File: src/auxkernels/Boundary2LDAux.C [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "Boundary2LDAux.h"
+
+registerMooseObject("DeerApp", Boundary2LDAux);
+
+defineLegacyParams(Boundary2LDAux);
+
+InputParameters Boundary2LDAux::validParams() {
```
Comment (reverendbedford) [inline src/auxkernels/Boundary2LDAux.C]: Capitalization, as per usual.

File: src/userobjects/Map2LDInterfaceMPRealVectorValue.C [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+#include "Map2LDInterfaceMPRealVectorValue.h"
+#include "MooseMesh.h"
+registerMooseObject("DeerApp", Map2LDInterfaceMPRealVectorValue);
+
+defineLegacyParams(Map2LDInterfaceMPRealVectorValue);
+
+InputParameters Map2LDInterfaceMPRealVectorValue::validParams() {
+  InputParameters params = Map2LDInterfaceMaterialPropertyUO::validParams();
+  params.addRequiredParam<unsigned int>("component",
+                                        "the copmponent of the real vector");
```
Comment (reverendbedford) [inline src/userobjects/Map2LDInterfaceMPRealVectorValue.C]: Spelling and capitalization

File: src/userobjects/Map2LDelem.C [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+Map2LDelem::Map2LDelem(const InputParameters &parameters)
+    : InterfaceUserObject(parameters),
+      ld_block_ids(_mesh.getSubdomainIDs(
+          getParam<std::vector<SubdomainName>>("ld_block_names"))) {}
+
+Map2LDelem::~Map2LDelem() {}
+
+void Map2LDelem::initialSetup() {
+  // define the boundary map nad retrieve element side and boundary_ID
```
Comment (reverendbedford) [inline src/userobjects/Map2LDelem.C]: Spelling

File: src/userobjects/Map2LDelem.C [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    }
+  }
+}
+
+std::pair<dof_id_type, unsigned int>
+Map2LDelem::getLDNeighbor(dof_id_type ld_elem) const {
+  auto elem_side_pair = _map_LD_with_elem_side.find(ld_elem);
+  if (elem_side_pair != _map_LD_with_elem_side.end()) {
+    return elem_side_pair->second;
+  } else
```
Comment (reverendbedford) [inline src/userobjects/Map2LDelem.C]: Use brackets.

File: src/userobjects/Map2LDelem.C [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+Map2LDelem::Map2LDelem(const InputParameters &parameters)
+    : InterfaceUserObject(parameters),
+      ld_block_ids(_mesh.getSubdomainIDs(
+          getParam<std::vector<SubdomainName>>("ld_block_names"))) {}
+
+Map2LDelem::~Map2LDelem() {}
+
+void Map2LDelem::initialSetup() {
```
Comment (reverendbedford) [inline src/userobjects/Map2LDelem.C]: This method is way too big.  Split up into logical parts (use private member functions if that makes sense).

**Comment by arovinelli** [comment]
Let's put his on hold until [14705](https://github.com/idaholab/moose/pull/14705) is not merged as It depends on it

**Pull request closed without merging**